src/ucp/core/ucp_worker.c

```diff
@@ -754,40 +760,53 @@ void ucp_worker_iface_event(int fd, void *arg)
 }
 
 static int ucp_worker_iface_find_better(ucp_worker_h worker,
-                                        ucp_worker_iface_t *wiface)
+                                        ucp_worker_iface_t *wiface,
+                                        ucp_rsc_index_t *better_index)
 {
     ucp_context_h ctx = worker->context;
-    ucp_rsc_index_t id;
+    ucp_rsc_index_t rsc_index;
     ucp_worker_iface_t *if_iter;
+    uint64_t test_flags;
+    double latency_iter, latency_cur;
+    float epsilon;
 
-    for (id = 0; id < ctx->num_tls; ++id) {
-        if_iter = &worker->ifaces[id];
+    test_flags = wiface->attr.cap.flags & ~(UCT_IFACE_FLAG_CONNECT_TO_IFACE |
+                                            UCT_IFACE_FLAG_CONNECT_TO_EP);
+
+    for (rsc_index = 0; rsc_index < ctx->num_tls; ++rsc_index) {
+        if_iter = &worker->ifaces[rsc_index];
 
         /* Need to check resources which belong to the same device only */
-        if ((ctx->tl_rscs[id].dev_index != ctx->tl_rscs[wiface->rsc_index].dev_index) ||
-            (if_iter->flags & UCP_WORKER_IFACE_FLAG_SUBOPTIMAL) ||
-            (id == wiface->rsc_index)) {
+        if ((ctx->tl_rscs[rsc_index].dev_index != ctx->tl_rscs[wiface->rsc_index].dev_index) ||
+            (if_iter->flags & UCP_WORKER_IFACE_FLAG_UNUSED) ||
+            (rsc_index == wiface->rsc_index)) {
             continue;
         }
 
         /* Check that another iface:
-         * 1. Supports all capabilities of the target iface (at least)
+         * 1. Supports all capabilities of the target iface (at least), except
+         *    ...CONNECT_TO... caps.
          * 2. Has the same or better performance charasteristics */
-        if (ucs_test_all_flags(if_iter->attr.cap.flags, wiface->attr.cap.flags) &&
-            (if_iter->attr.overhead          <= wiface->attr.overhead) &&
-            (if_iter->attr.latency.overhead  <= wiface->attr.latency.overhead) &&
-            (if_iter->attr.latency.growth    <= wiface->attr.latency.growth) &&
-            (if_iter->attr.bandwidth         >= wiface->attr.bandwidth) &&
-            (if_iter->attr.priority          >= wiface->attr.priority)) {
-
-            /* Do not check this iface anymore, because better one exists.
-             * It helps to avoid the case when two interfaces with the same caps
-             * and performance exclude each other. */
-            wiface->flags |= UCP_WORKER_IFACE_FLAG_SUBOPTIMAL;
-            return 1;
+        if (ucs_test_all_flags(if_iter->attr.cap.flags, test_flags) &&
+            (if_iter->attr.overhead  <= wiface->attr.overhead)      &&
+            (if_iter->attr.bandwidth >= wiface->attr.bandwidth)     &&
+            (if_iter->attr.priority  >= wiface->attr.priority)) {
```

Comment:
minor: in order to avoid more 'if' nesting level, maybe inverse the condition and do 'continue'.

Reply:
not sure I get it: currently in most cases we should fail on the very first condition checking capability bits, then all other conditions will not be checked. If everything, but latency matches, we will have to calculate latency funcs anyway

Reply:
just suggested to reduce nesting level to make it easier to read, not for performance (therefore it's 'minor')